### DISCUSSION TITLE
QQ - Intelligent Octopus "peak" hours

### DISCUSSION THREAD

**Discussion by Leventionz** [BottlecapDave/HomeAssistant-OctopusEnergy#407] [Q&A]
Hi

If IO schedules a charge during what would otherwise be a peak window, does the entire household load benefit from the reduced price (7.5p currently) or just the car?

I noted that the current cost sensor seems to suggest the price has gone down to 7.5p but when I look at today's schedule, that's quite a chunk of the peak period at the lower price, is that right or is it only applicable to the car?

Apologies, appreciate this isn't strictly related to this integration but wasn't certain if it was this integrations current rate sensor misbehaving and skewing the real price or my misunderstanding.

![aqptp2p1](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/45cfa23f-9364-4cd0-9458-6c571a3d87e1)

Thanks

**Comment by elatechi** [comment]
The proposed slots are not limited to car charging but all household consumption.  please note that as soon as you unplug the car, the slot will be removed and revert to peak hour rate.

**Reply by Leventionz** [reply]
Thanks, that is what I assumed, something isn't right then...

My car was plugged in between 9.30am and 12pm today, and the Intelligent Dispatching Sensor was ON.
I have a node-red script which checks this sensor and if needed will pull energy from the grid; it was definitely a 7.5p period.
I even have an additional check on the current price sensor is 7.5p in addition to the Intelligent Dispatching sensor before I charge the battery to prevent the issue I am about to describe.

You can see from the graph below the battery charged correctly based on this logic:

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/50e8f2f4-a2f5-40ec-b002-a8cd95a0b50f)

And below here is the Current Accumulative Cost for the same period:

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/cdb225d2-a199-4442-b1e4-93a22af1721b)

At first I see a linear increase as if it was only 7.5p up until 11.35 when my battery hit it's target 80% charge. 
Then at 12:01 the cost jumps from £2.41 to £4.26.

Upon further investigation I can see the rate for the entire period shows as 29.5554p but that definitely was not the case. It is as if the sensor has been retrospectively updated post 12pm. What makes even less sense is the cumulative cost during this period doesn't even at 29.5554p add up to the difference between £2.41 and £4.26, it would be £3.93.

- from: '2023-09-19T09:30:00+00:00'
  to: '2023-09-19T10:00:00+00:00'
  rate: 29.5554p
  consumption: 1.944 kWh
  consumption_raw: 1.944
  cost: £0.57
  cost_raw: 0.57
- from: '2023-09-19T10:00:00+00:00'
  to: '2023-09-19T10:30:00+00:00'
  rate: 29.5554p
  consumption: 1.906 kWh
  consumption_raw: 1.906
  cost: £0.56
  cost_raw: 0.56
- from: '2023-09-19T10:30:00+00:00'
  to: '2023-09-19T11:00:00+00:00'
  rate: 29.5554p
  consumption: 1.246 kWh
  consumption_raw: 1.246
  cost: £0.37
  cost_raw: 0.37
- from: '2023-09-19T11:00:00+00:00'
  to: '2023-09-19T11:30:00+00:00'
  rate: 29.5554p
  consumption: 0.038 kWh
  consumption_raw: 0.038
  cost: £0.01
  cost_raw: 0.01
- from: '2023-09-19T11:30:00+00:00'
  to: '2023-09-19T12:00:00+00:00'`
  
Further confirmation that Intelligent Dispatching was On during the time period:-
   
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/fa83c8f8-8f55-4255-a068-6133a48775b5)

Any ideas why this may have happened as it looks as though it reverted back to peak pricing retrospectively or is there an issue with the sensor?

Thank you

**Reply by BottlecapDave** [reply]
Hello and sorry that you're having issues with the integration. Is this an issue with the `previous accumulative cost` sensor or the `current accumulative cost` sensor?

> Upon further investigation I can see the rate for the entire period shows as 29.5554p but that definitely was not the case. It is as if the sensor has been retrospectively updated post 12pm.

The rates are retrieved from the APIs, but these alone only include the base rates (e.g peek and off peak). I then adjust these rates based on the pending and completed dispatches with either an undefined source or a source of `smart-charge`. What I'm not sure  is what this data looks like _during_ a smart charge - does the period stay in the pending list, does it move to completed, or does it disappear and then re-appear in the completed list once completed. I'm also not sure if OE put a slot in the pending list, if it's guaranteed or is there a chance that a period in the pending is never executed and moved to the completed list.

I'm not personally on intelligent, so these features rely on data provided by the community. If this is an issue, it would be benefical to raise a bug. When you do, it will help if you (or someone else) can turn on [debugging mode](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/faq.md#how-do-i-increase-the-logs-for-the-integration) an hour before this adhoc charging period starts, and then turn it off an hour after it's finished and provide the logs with the raised issue.

> What makes even less sense is the cumulative cost during this period doesn't even at 29.5554p add up to the difference between £2.41 and £4.26, it would be £3.93.

The total values for the sensors include the standing charge as well. There's an attribute which includes the value minus the standing charge. Does this account for the discrepancy?

**Reply by Leventionz** [reply]
Thank you for the detailed reply Dave.

I am using the `current accumulative cost` sensor to summarise today's cost (the £4.21 figure you see) however the raw data I provided above showing how it was 29.5554 during the AM periods I obtained from the `current rate` sensor, I have just checked the `previous accumulative cost` sensor and I cannot see any data relating to the 19th in sensors attributes yet so cannot answer about the standing charge question until this populates.

I can only see for yesterday in the `binary_sensor.octopus_energy_intelligent_dispatching` sensor:-

"Completed dispatches
- start: '2023-09-19T14:30:00+00:00'
  end: '2023-09-19T15:00:00+00:00'
  charge_in_kwh: -0.26
  source: null"

Just for clarity I did not charge my vehicle at all yesterday, it was connected to my wallbox to obtain Intelligent Dispatching (ID) times only.

------

I have just plugged the car in and it immediately wanted to charge, on this occasion I do need to charge the vehicle. I have enabled debug logging and can see the following planned dispatches:-

`Planned dispatches
- start: '2023-09-20T07:52:12+00:00'
  end: '2023-09-20T11:00:00+00:00'
  charge_in_kwh: -23.16
  source: smart-charge
- start: '2023-09-20T14:00:00+00:00'
  end: '2023-09-20T14:30:00+00:00'
  charge_in_kwh: -0.58
  source: smart-charge
- start: '2023-09-20T14:30:00+00:00'
  end: '2023-09-20T15:00:00+00:00'
  charge_in_kwh: -3.7
  source: smart-charge`

My suspicion is that the issue will not present itself when I am actually wanting to charge the vehicle, rather when I plug the vehicle in but restrict the charge target using my car's App to prevent a "smart charge". If the vehicle is already at a satisfactory level (but to take advantage of cheap household energy). Maybe if nothing goes through my car charger and OE are able determine this and revert the period?
I will provide logs when charging the vehicle and ID is active and then a separate set of logs when not charging and ID is active.

Thank for the support.

**Comment by Leventionz** [comment]
> What I'm not sure is what this data looks like _during_ a smart charge - does the period stay in the pending list, does it move to completed, or does it disappear and then re-appear in the completed list once completed.

Charging now and it remains in planned dispatches.

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/6fb10105-9dbf-4f43-b8a6-7a1a82979c42)

Will update once this first period finishes.

**Comment by Leventionz** [comment]
Car plugged in the entire time, looks like 3 half hour periods moved to completed when the car was charging, and the others have disappeared (9am-11am UTC, 10-12pm BST).

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/e166d7a6-086b-4f46-a6d6-5d803994b497)

I can see my car did charge from 8am UTC to 9am UTC which correlates with the completed sessions; it then stopped charging but the Dispatching Sensor remains ON the entire period. I will provide the debug logs for you to review as until I get my bill through at the end of the month I cannot see if this is expected behaviour.

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/e3154e07-4982-4f65-a7a9-6bc0a93df911)

**Reply by BottlecapDave** [reply]
Thanks for the update. The intelligent data is requested every minute. The completed dispatches get dropped by OE quite quickly, so the integration captures a local list merging the local list with the data returned by OE up to 3 days worth of data. The sensor is meant to update frequently, but it might only update when new data is retrieved. I'll see if there is anything that can be done to improve the accuracy of the intelligent sensor.

**Comment by Leventionz** [comment]
Still waiting for my bill so will update on that shortly to see if it matters if the car is charging within a 30 minute period or if just being plugged in counts; from my testing so far I think the car needs to receive a charge within an Intelligent Dispatching 30 minute time period for it to appear in 'Completed Dispatches' and the household receive the benefit of the off-peak price period.

On a related note, I was going to create a new topic but as the theme is similar I noted some unusual behaviour this morning.

Intelligent Dispatching Sensor was 'Off' yet I can see 'Completed Dispatches' (UTC times so +1 for BST)

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/24567397/5a5f0242-b1d8-4b51-b3b4-9a9abdf34225)

- start: '2023-09-26T05:30:00+00:00'
  end: '2023-09-26T06:00:00+00:00'
  charge_in_kwh: -1.95
  source: null
- start: '2023-09-26T05:00:00+00:00'
  end: '2023-09-26T05:30:00+00:00'
  charge_in_kwh: -0.1
  source: null
- start: '2023-09-26T06:30:00+00:00'
  end: '2023-09-26T07:00:00+00:00'
  charge_in_kwh: -0.38
  source: null
- start: '2023-09-26T06:00:00+00:00'
  end: '2023-09-26T06:30:00+00:00'
  charge_in_kwh: -2.13
  source: null

As per my observations stated above it is my understanding that 'Completed Dispatches' only populates when Intelligent Dispatching is enabled as that forms part of the calculation for an off-peak price period? If you would like me to raise a separate bug for this I can do, it's the first time I noticed it; but wanted to check my understanding of the sensor before I did so to save wasting your time.

Another minor one that I do not understand is why the source always shows as null; this happens all the time. In fact every time the car is charged I never see the source as smart-charge.

Thanks

**Reply by BottlecapDave** [reply]
> As per my observations stated above it is my understanding that 'Completed Dispatches' only populates when Intelligent Dispatching is enabled as that forms part of the calculation for an off-peak price period?

As per the [docs](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/sensors/intelligent.md#is-dispatching), the dispatching sensor should turn on during off peak periods and at any point a pending or completed dispatch is active. Based on your [comment](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/discussions/407#discussioncomment-7058389) it sounded like dispatches stayed in pending until they were complete? Pending and completed dispatches are then used to retrofit current/previous consumption cost sensors, so if this data changes, then these sensors could adjust during the day. Previous consumption cost should be relatively stable, as I wouldn't expect the completed dispatches to change for yesterday.

There is nothing in the API to my knowledge that states a charge is currently being done, so I have to "best guess", which to my knowledge could be wrong, as a pending dispatch isn't guaranteed based on your comments. This sensor's aim was to mimic the behaviour of another intelligent integration which was folded into this integration as they have little time for maintainance.

Was the sensor off during an active dispatch period?

> Another minor one that I do not understand is why the source always shows as null; this happens all the time. In fact every time the car is charged I never see the source as smart-charge.

For some reason OE don't provide the original source in the completed dispatches, despite the field being present. This is the same in other integrations (see first screenshot of https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/issues/409). Not a lot I can do about this I'm afraid, unless I start cross comparing pending with completed, but that can get messy as they don't aways align up from my understanding.

**Reply by Leventionz** [reply]
The Dispatching Sensor was 'off' but I still got Completed Dispatches which implies it was actually on for this period. Unfortunately my bill stops on 25th so I cannot confirm if I was billed off-peak or not.
